lib/Dialect/LLHD/IR/LLHDDialect.cpp

```diff
@@ -84,6 +84,31 @@ static Type parseSigType(DialectAsmParser &parser) {
   return SigType::get(underlyingType);
 }
 
+/// Parse an array type.
+/// Syntax: array ::= !llhd.array<{length}x{elementType}>
+static Type parseArrayType(DialectAsmParser &parser) {
+  Type elementType;
+  SmallVector<int64_t, 1> length;
+  if (parser.parseLess() || parser.parseDimensionList(length, false) ||
+      parser.parseType(elementType) || parser.parseGreater()) {
+    parser.emitError(parser.getCurrentLocation(),
+                     "No array type found. Array needs a length and an element "
+                     "type.");
+    return nullptr;
+  }
+  if (length.size() != 1) {
+    parser.emitError(parser.getCurrentLocation(),
+                     "Array must have exactly one dimension");
+    return nullptr;
+  }
+  if (length[0] < 0) {
```

Comment:
Do you think this check could be moved to the `verifyConstructionInvariants` method together with the `length == 0` check?

Reply:
`ArrayType::get` takes an unsigned integer as its first argument, so `length[0]` will be implicitly cast to an unsigned by converting negative numbers to high positive ones. As `verifyConstructionInvariants` is called after this conversion, it wouldn't detect this case. However, this test is probably not needed at all as `parseDimensionList` seems to ignore negative numbers and as a result parsing the type will fail. (Side note: `parseDimensionList` would return -1 for dynamic dimensions, but I turned them off with the second argument). So I think it's save to completely remove it. Thanks!

Reply:
Sounds good!